File: 프로그래머스/2/92341. 주차 요금 계산/주차 요금 계산.java
import java.util.*;
import java.util.stream.*;

class Solution {
    private static class Car {
        String status;
        int totalTime;
        int inTime;

        public Car(String status, int totalTime, int inTime) {
            this.status = status;
            this.totalTime = totalTime;
            this.inTime = inTime;
        }
    }

    public int[] solution(int[] fees, String[] records) {
        Map<String, Car> history = new HashMap<>();
        for (String record : records) {
            String[] tokens = record.split(" ");
            String time = tokens[0];
            String carNum = tokens[1];
            String status = tokens[2];

            if (status.equals("IN")) {
                if (!history.containsKey(carNum)) {
                    int inTime = convert(time);
                    history.put(carNum, new Car(status, 0, inTime));
                } else {
                    Car car = history.get(carNum);
                    car.status = status;
                    car.inTime = convert(time);
                    history.put(carNum, car);
                }
            } else {
                if (history.containsKey(carNum)) {
                    Car car = history.get(carNum);
                    int outTime = convert(time);
                    car.status = status;
                    car.totalTime += outTime - car.inTime;
                }
            }
        }

        for (String carNum : history.keySet()) {
            Car car = history.get(carNum);
            if (car.status.equals("IN")) {
                car.status = "OUT";
                car.totalTime += convert("23:59") - car.inTime;
            }
        }

        List<Integer> result = new ArrayList<>();
        List<String> carNums = new ArrayList<>(history.keySet());
        Collections.sort(carNums);

        for (String carNum : carNums) {
            Car car = history.get(carNum);
            int price = 0;
            if (car.totalTime <= fees[0]) {
                price = fees[1];
            } else {
                price = fees[1] + (int) Math.ceil((double) (car.totalTime - fees[0]) / fees[2]) * fees[3];
            }
            result.add(price);
        }
        return result.stream().mapToInt(Integer::intValue).toArray();
    }

    private int convert(String time) {
        String[] tokens = time.split(":");
        return Integer.parseInt(tokens[0]) * 60 + Integer.parseInt(tokens[1]);
    }
}

File: 프로그래머스/2/92341. 주차 요금 계산/주차 요금 계산.cpp
#include <string>
#include <vector>
#include <map>
#include <sstream>
#include <iostream>
using namespace std;

struct Car
{
    int num;
    int totalTime;
    int inTime;
    
    Car(int n, int tt, int it)
    {
        this->num = n;
        this->totalTime = tt;
        this->inTime =it;
    }
};

vector<string> Split(string s, char del)
{
    vector<string> answer;
    stringstream ss (s);
    string temp;
    
    while(getline(ss, temp, del))
    {
        answer.push_back(temp);
    }
    
    return answer;
}

int Transform(string t)
{
    int hour = stoi(t.substr(0, 2));
    int min = stoi(t.substr(3, 2));
    
    return hour * 60 + min;
}

vector<int> solution(vector<int> fees, vector<string> records) {
    vector<int> answer;
    map<int, Car*> m;
    
    for(int i = 0; i < records.size(); ++i)
    {
        vector<string> data = Split(records[i], ' ');
        int time = Transform(data[0]);
        int carNum = stoi(data[1]);
        string cmd = data[2];
        
        if (cmd == "IN")
        {
            if (m.find(carNum) == m.end())
            {
                m.insert(make_pair(carNum, new Car(carNum, 0, time)));
            }
            else
            {
                m[carNum]->inTime = time;
            }
        }
        else
        {
            m[carNum]->totalTime += time - m[carNum]->inTime;
            m[carNum]->inTime = -1;
            
        }
    }
    
    for(auto car : m)
    {
        Car* c = car.second;
        
        if (c->inTime != -1)
        {
            c->totalTime += Transform("23:59") - c->inTime;
        }
        
        int totalTime = c->totalTime;
        int cost = fees[1];
        totalTime -= fees[0];
        
        if (totalTime > 0)
        {
            cost += totalTime / fees[2] * fees[3];
        
            if (totalTime % fees[2] != 0)
            {
                cost += fees[3];
            }
        }
        
        answer.push_back(cost);
    }
    
    return answer;
}